### PR TITLE
Gamemode Entities

### PR DESCRIPTION
**Pull request by DevulTj**
I've written a loose implementation for gamemode entities. This means that anything using IGamemodeEntity will register itself to a gamemode so the gamemode can access it.

They use target names (perhaps we should change this though) - so your entity's name should match the target name that you give your gamemode entity.

You can then either set up tags for your gamemode entities in Hammer: ```BaseGamemode.GetEntitiesWithTag( string tag )```

Or filter by type: ```BaseGamemode.GetEntitiesOfType( string tag = null )```

The idea is that we can set up entire gamemodes in Hammer. Ex: Set up camera positions using specific entities, and set tags for context.

### CODE REVIEW

File: code/Base/GamemodeSystem/Entities/BaseGamemode.TargetEntities.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+{
+	public List<Entity> Entities { get; protected set; } = new();
+
+	public IEnumerable<Entity> GetEntitiesWithTag( string tag )
+	{
+		return Entities.Where( x => x.Tags.Has( tag ) );
+	}
+	public IEnumerable<T> GetEntitiesOfType<T>( string tag = null )
+	{
+		if ( tag is not null )
```
Comment (DoctorGurke) [inline code/Base/GamemodeSystem/Entities/BaseGamemode.TargetEntities.cs:13]: The logic here seems to be the wrong way around?

File: code/Base/GamemodeSystem/Entities/BaseGamemode.TargetEntities.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+			if ( entity == this )
+				continue;
+
+			Log.Debug( $"Found gamemode entity for {Name} - {entity.GetType()}" );
+
+			if ( entity is IGamemodeEntity gamemodeEntity )
+			{
+				Entities.Add( entity );
+				gamemodeEntity.Gamemode = this;
+				entity.Parent = this;
```
Comment (DoctorGurke) [inline code/Base/GamemodeSystem/Entities/BaseGamemode.TargetEntities.cs:40]: good thinking with parenting them. We'll have to make sure map cleanup works properly along the line by parenting stuff to the game and/or using a custom cleanup filter. But that's just a random sidenote

File: code/Base/GamemodeSystem/Entities/BaseGamemode.TargetEntities.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+		{
+			return GetEntitiesWithTag( tag )
+				.OfType<T>();
+		}
+	}
+
+	[Event.Entity.PostSpawn]
+	protected virtual void GatherTargetEntities()
+	{
+		var ents = FindAllByName( Name );
```
Comment (DoctorGurke) [inline code/Base/GamemodeSystem/Entities/BaseGamemode.TargetEntities.cs:27]: for the sake of I/O targetnames we should probably use a property on GamemodePointEntity that links them to the gamemode instead, similar to how we do it with the orbs.

File: code/Base/GamemodeSystem/Entities/BaseGamemode.TargetEntities.cs (hunk 1 above)
Comment (DevulTj) [reply code/Base/GamemodeSystem/Entities/BaseGamemode.TargetEntities.cs:13]: Yeah, whoops

**Comment by DevulTj** [comment]
I've refactored some of it. It's now using an explicit property - and I've removed IGamemodeEntity as it was causing too many problems than I was trying to solve with it.

File: code/Base/GamemodeSystem/Entities/BaseGamemode.TargetEntities.cs (hunk 2 above)
Comment (DevulTj) [reply code/Base/GamemodeSystem/Entities/BaseGamemode.TargetEntities.cs]: Yeah!

File: code/Base/GamemodeSystem/Entities/BaseGamemode.TargetEntities.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
+namespace Sports;
+
+public partial class BaseGamemode
+{
+	public List<GamemodePointEntity> Entities { get; protected set; } = new();
+
+	public IEnumerable<Entity> GetEntitiesWithTag( string tag )
+	{
+		return Entities.Where( x => x.Tags.Has( tag ) );
+	}
```
Comment (DoctorGurke) [inline code/Base/GamemodeSystem/Entities/BaseGamemode.TargetEntities.cs:10]: spacing here

File: code/Base/GamemodeSystem/Entities/BaseGamemode.TargetEntities.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+				.OfType<T>();
+		}
+	}
+
+	[Event.Entity.PostSpawn]
+	protected virtual void GatherTargetEntities()
+	{
+		var ents = All.OfType<GamemodePointEntity>();
+
+		Log.Info( $"[{Name}] Finding gamemode entities, count: {ents.Count()}" );
```
Comment (DoctorGurke) [inline code/Base/GamemodeSystem/Entities/BaseGamemode.TargetEntities.cs:29]: This should probably be Log.Debug

File: code/Base/GamemodeSystem/Entities/BaseGamemode.TargetEntities.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+		foreach ( var gamemodeEntity in ents )
+		{
+			if ( gamemodeEntity.GamemodeId.ToLower() == Name.ToLower() )
+			{
+				Log.Info( $"	> Matched GamemodePointEntity for {Name}" );
```
Comment (DoctorGurke) [inline code/Base/GamemodeSystem/Entities/BaseGamemode.TargetEntities.cs:35]: This too

**Review by DoctorGurke — CHANGES_REQUESTED** [review]

File: code/Base/GamemodeSystem/Entities/BaseGamemode.TargetEntities.cs (hunk 6 above)
Comment (DevulTj) [reply code/Base/GamemodeSystem/Entities/BaseGamemode.TargetEntities.cs:35]: This one is intended so it looks like this
sbox_Tz2JySglv7

Comment (DevulTj) [reply code/Base/GamemodeSystem/Entities/BaseGamemode.TargetEntities.cs:35]: Oh you meant log.debug mb

**Pull request merged**